libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/incremental/ClasspathChangesComputer.kt

```diff
@@ -65,20 +68,37 @@ object ClasspathChangesComputer {
         for (key in current.classSnapshots.keys) {
             val currentSnapshot = current.classSnapshots[key]!!
             val previousSnapshot = previous.classSnapshots[key] ?: return Failure.AddedRemovedClasses
-            val result = collectClassChanges(currentSnapshot, previousSnapshot, changesCollector)
-            if (result is Failure) {
-                return result
+            if (currentSnapshot !is KotlinClassSnapshot || previousSnapshot !is KotlinClassSnapshot) {
+                return Failure.NotYetImplemented
             }
+            // TODO: Store results in changesCollector
+            collectKotlinClassChanges(currentSnapshot, previousSnapshot)
         }
         return Success
     }
 
-    private fun collectClassChanges(
-        current: ClassSnapshot,
-        previous: ClassSnapshot,
-        changesCollector: ChangesCollector
-    ): ChangesCollectorResult {
-        return Failure.NotYetImplemented
+    @TestOnly
+    internal fun collectKotlinClassChanges(current: KotlinClassSnapshot, previous: KotlinClassSnapshot): DirtyData {
+        val workingDir = FileUtil.createTempDirectory(this::class.java.simpleName, "WorkingDir", /* deleteOnExit */ true)
+        val incrementalJvmCache = IncrementalJvmCache(workingDir, /* targetOutputDir */ null, FileToCanonicalPathConverter)
```

Comment:
Can IncrementalJvmCache store all classes from `ClasspathSnapshot`, and then we can create one in `getChanges` and pass it around to load/store class info? Right now, we create an instance per class. It is mostly lazy, but there is some overhead.

Reply:
Yes I planned to do that too, added a TODO. Right now my plan is:
 - Snapshot and diff correctly for a Kotlin class (this PR)
 - Snapshot and diff correctly for a Java class (next PR)
 - Connect everything up to get a working version and update integration tests, add unit tests (PR after that)